File: 0053-maximum-subarray/0053-maximum-subarray.py
class Solution:
    def maxSubArray(self, nums: List[int]) -> int:

        maxVal, totalMax = nums[0], nums[0]
        for num in nums[1:]:
            maxVal = max(num, maxVal+num)
            totalMax = max(maxVal, totalMax)
        return totalMax
        
'''
        
        -2, 1,-3, 4,-1, 2, 1,-5, 4

        start with first index
        1 vs 1
        -2 vs -3
        -2 vs 4
        3 vs -1
        5 vs 2
        6 vs -5

        continuous subarray with largest sum


'''        

File: 0053-maximum-subarray/0053-maximum-subarray.cpp
class Solution {
public:
    int maxSubArray(vector<int>& nums) {
        if(nums.size() == 0) {
            return -1;
        }

        int curr=nums[0], ans=nums[0];

        for(int i=1; i<nums.size(); i++) {
            curr = max(nums[i], nums[i]+curr);
            ans = max(curr, ans);
        }
        return ans;
    }
};